### PR TITLE
Ideal nav LHP

### PR DESCRIPTION
**Pull request by WojtekBoman**
### Details
This PR introduces the Left Hand Pane to the app. It's implemented analogously to the Right Hand Pane. 
The most important changes: 
- Search screen has been moved to the LHP
- Added the configuration for the LHP based on the RHP configuration with some adjustments (LHP animation is reverted)
- linkTo function has been updated to handle dismissing LHP
- Referral has been removed from the SearchPage

### Fixed Issues

$ https://github.com/Expensify/App/issues/32939
PROPOSAL: described in the design doc here: https://docs.google.com/document/d/1VnEf2ThbbiQj0kx9KnAihF0DWf50s2Bn_MztEyEwtjg/edit#heading=h.r39uu4rszpv

### Tests

1. Log in to the app.
2. Press on the search icon at the top of the sidebar.
3. Check if the search screen is opened on the left side of the screen.

- [x] Verify that no errors appear in the JS console

### Offline tests

1. Log in to the app.
2. Press on the search icon at the top of the sidebar.
3. Check if the search screen is opened on the left side of the screen.

### QA Steps

1. Log in to the app.
2. Press on the search icon at the top of the sidebar.
3. Check if the search screen is opened on the left side of the screen.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/47774969/926b60ec-b791-41b7-a882-1094c2341d7a

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/47774969/769f581d-72b3-42de-ad6a-85f9cf53a656

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/47774969/4702e767-a0b9-4f23-9c82-bad3ddcdfa53

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/47774969/3d25dee1-3dc2-410b-9240-53dbc5f7858c

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/47774969/e0bc9534-25ee-446c-945a-d3e07a2d358c

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/47774969/077d2617-c7bb-4ac1-a20d-99f2c03b8241

</details>

### CODE REVIEW

File: src/libs/Navigation/linkTo.ts [hunk 1]
```diff
@@ -89,12 +97,15 @@ export default function linkTo(navigation: NavigationContainerRef<RootStackParam
             action.type = CONST.NAVIGATION.ACTION_TYPE.REPLACE;
 
             // If this action is navigating to the RightModalNavigator and the last route on the root navigator is not RightModalNavigator then push
```
Comment (s77rt) [inline src/libs/Navigation/linkTo.ts:99]: Update comment

**Comment by hayata-suenaga** [comment]
@WojtekBoman is this supposed to merged into the main branch?

File: src/libs/Navigation/linkTo.ts [hunk 2]
```diff
@@ -89,12 +97,15 @@ export default function linkTo(navigation: NavigationContainerRef<RootStackParam
...
-        } else if (action.payload.name === NAVIGATORS.RIGHT_MODAL_NAVIGATOR && rootState?.routes?.at(-1)?.name !== NAVIGATORS.RIGHT_MODAL_NAVIGATOR) {
+        } else if (isModalNavigator(action.payload.name) && !isTargetNavigatorOnTop) {
+            if (isModalNavigator(topRouteName)) {
+                dismissModal('', navigation);
+            }
```
Comment (s77rt) [inline src/libs/Navigation/linkTo.ts:103]: Can you please clarify why is this needed?

File: src/libs/Navigation/Navigation.ts
Comment (s77rt) [inline src/libs/Navigation/Navigation.ts:1]: `goBack` function contains logic for the RHP (isFirstRouteInNavigator). We should account for the LHP too

**Review by s77rt — CHANGES_REQUESTED** [review]

**Comment by s77rt** [comment]
NAB: `CustomRouter` contains logic to extract the report id from the params to show the correct report id in the central panel. We should probably do the same for the LHP.

File: src/libs/Navigation/linkTo.ts (hunk 2 above)
Comment (WojtekBoman) [reply src/libs/Navigation/linkTo.ts:103]: It's necessary to dismiss the already opened modal when we navigate to ModalNavigator. Before I removed the referral from the SearchPage, there had been possibility to navigate from LHP to RHP. In this case LHP should be dismissed before RHP is opened. It's the additional check for that.

File: src/libs/Navigation/Navigation.ts [hunk 3]
```diff
@@ -56,7 +58,7 @@ function getActiveRouteIndex(stateOrRoute: StateOrRoute, index?: number): number
         return getActiveRouteIndex(childActiveRoute, stateOrRoute.state.index ?? 0);
     }
 
-    if ('name' in stateOrRoute && stateOrRoute.name === NAVIGATORS.RIGHT_MODAL_NAVIGATOR) {
+    if ('name' in stateOrRoute && (stateOrRoute.name === NAVIGATORS.RIGHT_MODAL_NAVIGATOR || stateOrRoute.name === NAVIGATORS.LEFT_MODAL_NAVIGATOR)) {
```
Comment (WojtekBoman) [inline src/libs/Navigation/Navigation.ts:61]: > `goBack` function contains logic for the RHP (isFirstRouteInNavigator). We should account for the LHP too

@s77rt I've modified this condition to check it

**Comment by mountiny** [comment]
@hayata-suenaga correct this should not be influencing the current code so it can go to `main`

File: src/libs/Navigation/linkTo.ts (hunk 2 above)
Comment (WojtekBoman) [reply src/libs/Navigation/linkTo.ts:103]: Furthermore, we need this check to properly handle keyboard shortcuts in the app. When we switch from the LHP to the RHP, the previous one should be closed. 

https://github.com/Expensify/App/assets/47774969/027cb2ac-63d7-4567-9d84-734dc1876790


https://github.com/Expensify/App/assets/47774969/b4117796-1f00-4c75-ad83-097f685d5f11

**Comment by WojtekBoman** [comment]
> NAB: `CustomRouter` contains logic to extract the report id from the RHP screen params to show the correct report in the central panel. We should probably do the same for the LHP.

You're right but currently we don't store any page connected with specific report in the LHP, so there is no reportID to extract

**Comment by mountiny** [comment]
@WojtekBoman The overlay seems to be weirdly animating here 
image

**Comment by WojtekBoman** [comment]
> @WojtekBoman The overlay seems to be weirdly animating here 

@mountiny Thank you for noticing this. I added the fix and now the animation should work fine 

https://github.com/Expensify/App/assets/47774969/faf098af-8f1a-46e9-b9d8-bdfd7bf5ef58

File: src/libs/Navigation/linkTo.ts (hunk 2 above)
Comment (s77rt) [reply src/libs/Navigation/linkTo.ts:103]: Makes sense!

**Comment by s77rt** [comment]
## Reviewer Checklist

- [X] I have verified the author checklist is complete (all boxes are checked off).
- [X] I verified the correct issue is linked in the `### Fixed Issues` section above
- [X] I verified testing steps are clear and they cover the changes made in this PR
    - [X] I verified the steps for local testing are in the `Tests` section
    - [X] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [X] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [X] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I verified tests pass on **all platforms** & I tested again on:
    - [X] Android: Native
    - [X] Android: mWeb Chrome
    - [X] iOS: Native
    - [X] iOS: mWeb Safari
    - [X] MacOS: Chrome / Safari
    - [X] MacOS: Desktop
- [X] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



https://github.com/Expensify/App/assets/16493223/0b12144a-a70f-4cfb-9578-5e6572593934


</details>

<details>
<summary>Android: mWeb Chrome</summary>



https://github.com/Expensify/App/assets/16493223/71c83612-29c3-4496-bc98-426725d04c35


</details>

<details>
<summary>iOS: Native</summary>


https://github.com/Expensify/App/assets/16493223/a8915ead-cd71-445e-8a57-061a4d09828a

</details>

<details>
<summary>iOS: mWeb Safari</summary>



https://github.com/Expensify/App/assets/16493223/2e9a91f4-f215-44f3-a5f4-b777d884cd99


</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



https://github.com/Expensify/App/assets/16493223/209f478a-5af5-49be-acfa-4f99a8b5732f


</details>

<details>
<summary>MacOS: Desktop</summary>



https://github.com/Expensify/App/assets/16493223/934635f3-8ef0-430f-ae7e-3aa9c00ee671


</details>

File: src/libs/Navigation/Navigation.ts (hunk 3 above)
Comment (s77rt) [reply src/libs/Navigation/Navigation.ts:61]: But that does not seem enough. In goBack there is another check for the root name

Comment (WojtekBoman) [reply src/libs/Navigation/Navigation.ts:61]: I've added the additional check

**Review by s77rt — APPROVED** [review]

File: src/libs/Navigation/AppNavigator/Navigators/Overlay.tsx [hunk 4]
```diff
@@ -9,15 +9,18 @@ import CONST from '@src/CONST';
 type OverlayProps = {
     /* Callback to close the modal */
     onPress: () => void;
+
+    /* Returns whether a modal is displayed on the left side of the screen. By default, the modal is displayed on the right */
+    isModalOnTheLeft?: boolean;
 };
 
-function Overlay({onPress}: OverlayProps) {
+function Overlay({onPress, isModalOnTheLeft = false}: OverlayProps) {
```
Comment (hayata-suenaga) [inline src/libs/Navigation/AppNavigator/Navigators/Overlay.tsx:17]: I think no need for a default value. `isModalOnTheLeft` will be undefined if not passed

File: src/libs/Navigation/AppNavigator/getRootNavigatorScreenOptions.ts [hunk 5]
```diff
@@ -15,6 +15,8 @@ const commonScreenOptions: StackNavigationOptions = {
     animationTypeForReplace: 'push',
 };
 
+const SLIDE_LEFT_OUTPUT_RANGE_MULTIPLIER = -1;
```
Comment (hayata-suenaga) [inline src/libs/Navigation/AppNavigator/getRootNavigatorScreenOptions.ts:18]: should we move this to the const file or can we keep this here?

File: src/libs/Navigation/AppNavigator/Navigators/Overlay.tsx (hunk 4 above)
Comment (mountiny) [reply src/libs/Navigation/AppNavigator/Navigators/Overlay.tsx:17]: I think its cleaner to defined the default value than not.

File: src/libs/Navigation/AppNavigator/getRootNavigatorScreenOptions.ts (hunk 5 above)
Comment (WojtekBoman) [reply src/libs/Navigation/AppNavigator/getRootNavigatorScreenOptions.ts:18]: I think this variable won't be used elsewhere in the application, it is specific to function modelCardStyleInterpolator which is only called in this file

File: src/libs/Navigation/AppNavigator/Navigators/Overlay.tsx (hunk 4 above)
Comment (WojtekBoman) [reply src/libs/Navigation/AppNavigator/Navigators/Overlay.tsx:17]: Value of isModalOnTheLeft should always be passed to the Overlay component. I've added the appropriate comment in the `OverlayProps` type with the information that we display the modal on the right side of the screen by default

**Review by mountiny — APPROVED** [review]
Thank you, changes look good to me and I have only noticed one issue that when thr page is narrower than 920px roughly the overlay effect is still a bit odd. I think thats not a blocker though
